joomla/joomla-cms
# [4.0] Controller classes improvement

Pull Request for Issue # .

### Summary of Changes
This PR contains improvement + clean up to controller class to support both namespace and none namespace components in Joomla 4 (backward compatible change). Below are the list of changes/improvements:

1. **Change to base Controller class**
- Allow inject Input data instead of coupling with application input
- Support creating namespace model/view classes
- Simplify getName() method (that method was implemented wrong from beginning, instead of returning name of controller, it return component name ,without com_ prefix), just use substr method instead of regex.
- Implement getControllerName() method to allow getting name of controller properly (as mentioned above, that should be the role of getName() method)

2. **Change to Form Controller class**
- Call getControllerName method (from base controller class) to get context instead of using regex
- Use toPlural method of \Joomla\String\Inflector class to detect view_list property instead of using a script found elsewhere in the internet.
- Remove some redundant else if

3. **Change to Admin Controller class**
- Use controller input instead of application input to get input data (this was fixed in staging but not updated to 4.0 branch for some reasons)
- Implement getModel method. That will help us not having to add blank method like this https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_content/controllers/articles.php#L121 in every admin controller

### Testing Instructions

1. Install Joomla 4.0
2. Apply patch
3. Try to navigate to different pages (maybe try to add. edit, delete, publish articles), make sure it is still working as before

I also tested with the namespace com_content component in this PR https://github.com/joomla/joomla-cms/pull/14370 and it worked well. Hard for end user to test it for now, thought

Need developers to review code and discuss, too
